## IT309 - Assignment 6  
## Tim Nguyen Due 11/14/2023 | Assignment 6 (AS6)

### CA4 (In-class assignment #4) and Assignment 6 (A6)  
   
This in-class assignment is meant to help you launch the code you need to write for Assignment #6.  This carries 10 points, but A6 carries 40, so the combined total is 50.  
  
Experiment with hashing by creating a hash table and hash function. Use these to hash a set of input values provided in a transaction file.  Vary the size of the table and details of the hashing function to see:

1. how the load factor and number of collisions varies
2. how well the hash function covers the limited hash table area
3. does including more or less of the input data as hash function input change the results?
4. are the entries spread out or clustered?
  
Steps for CA4 (at least steps 1 - 3 below) and A6 (the rest of the steps):  
(1) Create the HT using a Python list, but with a fixed capacity.  
(2) HT slots are 'None' if unoccupied, or have data in them  
(3) add code to read the transaction file, create a transaction list of lists, with each sublist being the single transaction [trans code, integer]   
(4) add code to the Hash function to convert the input data into a number that fits within the size of the hash table  
(5) Do NOT use any built-in hash function or one downloaded - you must create your own        
(6) (a) add code to read the input and (b) populate the HT and handle collisions    
(7) Collision handling can be by probing or by chaining, which can be done using a list    
(8) After running the input data through the hashing process and filling in the HT, at the very least measure how much is occupied, the load factor, and the number of collisions. Include other measures that you can think of.      
(9) After trying different combinations of HT size and hash function specifics, compare the resulting measurements and provide a short written analysis.
   
Complete as much of this CA4 assignment as possible in class, at least through (3) and the creation of one hash function in (4).
Then upload completed code, including execution run(s), to the BB assignment site for CA4.   


### Steps (1) and (2) Create the Hash Table HT

In [12]:
# Basic Hash table set-up
HTcapacity = 98           # Allocate 97 slots in the Hash Table - can be varied
HT = [None]*HTcapacity    # 'HT' is the hash table data structure  
multFactor = 137          # multiplication factor used by Hash function - can be varied
print(len(HT))            # Verify HT capacity

98


### Step (3) Read the input file and convert the 'I' records to a list

In [13]:
fname = 'AS6input.txt'  # Make this the name of your local file
f = open(fname, 'r')
a6input = f.readlines()
f.close()
print(len(a6input))
print(a6input[0], '   ', a6input[-1])
print(a6input[80])

105
I,Rizzo,Mike,1962,M,PhEd
     F,Sagan,Carl,1938,M,Astronomy
I,Rodriguez,Hector,1987,M,ENGR



In [38]:
inlists = [ ]        # Create a list-of-lists for the input transactions
# insert code here to split the a6input lines into the list-of-lists
for line in a6input:
    transaction_line = line.strip().split(",")
    if transaction_line[0] =="I":  
        inlists.append(transaction_line)
    elif transaction_line[0] =="F":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            find_count += 1
    elif transaction_line[0] =="R":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            HT[hash_val].remove(transaction_line[1:])
            remove_count += 1

### Steps (4) and (5) - Add code to the Hash Function
### 'Hash' function performs the computation of the hash code and compression to the table size.
### Do NOT use any built-in hash functions or something you download. Create this yourself. 


In [37]:
def Hash (inline, mFactor = 1123):
    """Hash function - takes an input line and returns the HT code. """
    # insert code to convert an input line string into a number that ranges from
    # 0 to the capacity of the hash table HT
    transact_line=''.join(inline)
    hash_num = 0
    for i in transact_line:
        hash_num += ord(i) * mFactor   #Turns string into int
    return hash_num % HTcapacity            #Returns remainder as hash


### Step (6)(a) Compute and print the hash values for each input line.  Ignore short lines that are mostly '\n' and nondata.

In [16]:
# This runs all the input through the hash function and prints te resulting number
# Code is provided - no need to rewrite this cell.
for l in inlists:    # inlists is created in Step (3), has the input to be hashed
    if len(l) > 3:
        print('Hash values for "I" input records: ', l[1], l[2], ' is ',Hash(l, mFactor = multFactor))

Hash values for "I" input records:  Rizzo Mike  is  63
Hash values for "I" input records:  Womack Mary  is  19
Hash values for "I" input records:  Smith Marv  is  6
Hash values for "I" input records:  Miller Samuel  is  31
Hash values for "I" input records:  Kelvin Katrina  is  53
Hash values for "I" input records:  Sherman Maria  is  90
Hash values for "I" input records:  Ali Hassan  is  1
Hash values for "I" input records:  Smith Mary  is  35
Hash values for "I" input records:  Miller Smuel  is  50
Hash values for "I" input records:  Adams Stuart  is  19
Hash values for "I" input records:  Lomack Marv  is  1
Hash values for "I" input records:  Strasburg Stephen  is  9
Hash values for "I" input records:  Thompson Creek  is  38
Hash values for "I" input records:  Bryson William  is  8
Hash values for "I" input records:  Eaton Adam  is  45
Hash values for "I" input records:  Jones Johann  is  94
Hash values for "I" input records:  Kevin Katherine  is  69
Hash values for "I" input record

### Steps (6)(b) and (7) Compute the hash values (again) and insert into the table.  Handle collisions by chaining.

In [34]:
# Recompute the has value of each line.  Store the entire input record in the HT 'list' slot.  Chain
#   collisions by making the chain a 'list-of'list'.
HT = [None] * HTcapacity         # Create the HT, even if done in Step (1)
for i in inlists:
    if i[0] == 'I':              # only handle insertions
    # insert code to calculate the hash and insert into the table
        hash_val = Hash(i, mFactor=multFactor)
    # also handle collisions - can use probing or chain by appending results into a 
    # that occupies the table slot
        if HT[hash_val] is None:
            HT[hash_val] = [i]
            print("Inserted into slot:", i)
        else:
            HT[hash_val].append(i)
            print("[Collision occured] Inserted into slot:", i)
            

Inserted into slot: ['I', 'Rizzo', 'Mike', '1962', 'M', 'PhEd']
Inserted into slot: ['I', 'Womack', 'Mary', '1972', 'F', 'Math']
Inserted into slot: ['I', 'Smith', 'Marv', '1978', 'M', 'CS']
Inserted into slot: ['I', 'Miller', 'Samuel', '1974', 'M', 'HIST']
Inserted into slot: ['I', 'Kelvin', 'Katrina', '1982', 'F', 'ARTH']
Inserted into slot: ['I', 'Sherman', 'Maria', '1992', 'F', 'IST']
Inserted into slot: ['I', 'Ali', 'Hassan', '1996', 'M', 'MUSC']
Inserted into slot: ['I', 'Smith', 'Mary', '1999', 'F', 'ENGR']
Inserted into slot: ['I', 'Miller', 'Smuel', '1976', 'M', 'HIST']
[Collision occured] Inserted into slot: ['I', 'Adams', 'Stuart', '2000', 'M', 'Food']
[Collision occured] Inserted into slot: ['I', 'Lomack', 'Marv', '1971', 'M', 'Mgt']
Inserted into slot: ['I', 'Strasburg', 'Stephen', '1988', 'M', 'PhEd']
Inserted into slot: ['I', 'Thompson', 'Creek', '2000', 'F', 'ACCT']
Inserted into slot: ['I', 'Bryson', 'William', '1997', 'M', 'ENGL']
Inserted into slot: ['I', 'Eaton', 'A

### Print the Hash Table (HT) contents - from 0 to 101

In [39]:
# Print the hash table contents - shows what's in the table
# Code provided - no need to recreate
for n in range(0, len(HT)):
    print('index #', n, ' = ', HT[n])

index # 0  =  None
index # 1  =  [['I', 'Ali', 'Hassan', '1996', 'M', 'MUSC'], ['I', 'Lomack', 'Marv', '1971', 'M', 'Mgt']]
index # 2  =  [['I', 'Zane', 'Frank', '1942', 'Chem']]
index # 3  =  [['I', 'Martinez', 'David', '1971', 'M', 'PhEd'], ['I', 'Warden', 'Jack', '1998', 'M', 'Math']]
index # 4  =  None
index # 5  =  None
index # 6  =  [['I', 'Smith', 'Marv', '1978', 'M', 'CS']]
index # 7  =  None
index # 8  =  [['I', 'Bryson', 'William', '1997', 'M', 'ENGL'], ['I', 'Principato', 'Geraldo', '1994', 'M', 'ENGL']]
index # 9  =  [['I', 'Strasburg', 'Stephen', '1988', 'M', 'PhEd']]
index # 10  =  [['I', 'Johnson', 'Katherine', '1930', 'F', 'Math']]
index # 11  =  None
index # 12  =  None
index # 13  =  [['I', 'Lockhorn', 'Leonard', '1992', 'M', 'CHEM']]
index # 14  =  None
index # 15  =  None
index # 16  =  None
index # 17  =  None
index # 18  =  None
index # 19  =  [['I', 'Womack', 'Mary', '1972', 'F', 'Math'], ['I', 'Adams', 'Stuart', '2000', 'M', 'Food'], ['I', 'Hardcastle', 'Morvena

### Step (8) Compute and print various performance statistics for the run.

In [58]:
empty = 0            # empty HT slots
occupied = 0         # HT slots occupied (not = None)
collisions = 0       # HT slots with collisions - more than 1 entry at the same place
items = 0            # entries in the table and collision chains
chains = 0           # entries in HT table slots with > 1 entry (i.e. entry is a collision slot)
# Iterate through the HT and calculate the vairous stats - some are above
# Code goes here...
for slot in HT:
    if slot is None:
        empty += 1
    else:
        occupied += 1
        if len(slot)>1:
            collisions += 1       #increment by 1 since len(slot)>1 = collision
            chains += len(slot)-1   #length of (chained) slot minus 1 (first element)
        items += len(slot)  #increment item if slot is occupied
load_factor = items / HTcapacity

#Hash table stats
print("Empty slots:", empty)
print("Occupied slots:", occupied)
print("Collisions:", collisions)
print("Total items:", items)
print("Chains:", chains)
print("Load Factor:", load_factor)


Empty slots: 80
Occupied slots: 18
Collisions: 14
Total items: 87
Chains: 69
Load Factor: 0.8877551020408163


### Step (9) Analysis - to be completed after experimenting

The hash function that I implemented in this assignment utilizes a given python library function ord(), which returns the ASCII unicode code (an integer) of a given character. The function first concatenates all strings in the parameter inline, going through the transaction line from text file "AS6input.txt" and turning every line into a single string for hashing. After the character (or key) is converted into an integer in the for loop, each are then multiplied my mFactor - which represents a constant that provides randomness and flexibility in the hash function. This in turn can affect the distribution of how keys are spread on the hash table and collissions. This step is repeated for every character in the string and essentially follows H1:(key -> integer) which converts a key (or character) into a integer (which also represents the character - vice versa) making it ready for part 2 of hashing, compression.

The compression part of the hash function follows the formula of H2: integers -> [0, N-1] where HTcapacity is N. It's fairly simple in its execution using the modulo of the H1 integer (our hash) and HTcapacity, which is the capacity of our hash table. The purpose of using modulo is to compress the bloated hash values into a more manageable number that lies within a specifc range. Since the first part of hashing requires mapping of the characters into integers, which in most cases wouldd produce a very long integer that is hard to work with. And since we're working with a fixed size slot - it is even more important to use compression to map values into slots. The main purpose is to have an even distribution in accordance to their values allowing for efficient use of memory to optimize operations such as insert, search, and remove.

Collisions occur when inputs have the same hash value which were handled by chaining together records - entries with the same values were stored in linked strings within slots of the hash table. The code basically checks if the slot is none, if it is then insert the record into this empty slot of the hash table and then print. If the slot is not empty, then appened the new record into the existing slot where a chain of strings is 'linked' into the same slot and print "Collision" to notify.

For the first test, I got:
Empty slots: 80
Occupied slots: 18
Collisions: 14
Total items: 87
Chains: 69
Load Factor: 0.88775510204

1) Selecting only a subset of input characters would change both collision and number slots as it would depend on which subset of characters were chosen. For instance, it could significantly impact collisions if 'selecting only a subset of numbers' would increase the uniformality of distribution leading to less collisions but possibly more slot occupation. This could happen the other way as well, if the subset of numbers were to decrease uniformality in the hash table, more collisions and more chaining would happen. In my experience testing the hash and variables, this would lead to a bigger load time as an increase in chains = decrease in effiency of the hash table. For example, test #4 (changing the HTCapacity to half) yields the results: Chains: 45 and Load Factor: 1.7755102040816326

Compared to the results of the first test, where the total number of chains was 29 and load factor was 0.888 - decrease of the slot capacity would result in increased load factors and an inefficiency in optimization.

2) Yes, changing H1 (hashing of keys to integers) would affect how collisions and open table slot's frequency of occuring. A very big and broad hash would mean a more evenly distributed use of slots, decreasing collisions of hashes that share the same value. However, this would increase the number of occupied slots, as shown below in test #3. The amount of occupied slots increased from 18 to 58 but the number of chains decreased from 69 to 29, while maintaining the same load factor. Test #3 For this test, I changed the variable of the mFactor by double to 2246 instead of 1123. Which changes H1 to a bigger hash
Empty slots: 40
Occupied slots: 58
Collisions: 24
Total items: 87
Chains: 29
Load Factor: 0.8877551020408163

5) If we were to reduce the size the size of the table capacity, it would mean fewer slots for hash values to go into - which would increase the number of collisions and chains. It may pose a problem as different keys may end up linked in the same spot more frequently... leading to bigger chains and higher load times as shown below in test #4


Consequently, if we were to increase the size of the table capacity (shown in test #5), we'll see that there is less chaining as more values are getting their own slot, reducing the amount of collisions. Here, we see increasing table size leads to a shorter load time for these sets of items. As shown in the formula where load factor = number of items / table size. If any of these variables were to change, it would inherently affect the efficciency of the table.







## Test #2 Other tests of using hashtable:

#### This is a test of the hash function I made from in-class assignemnt 4 which doesn't really follow the compression function of H2: integers -> (0, N-1) rule for returning the hash, rather it returns the converted key integer modulo 25

In [59]:
# Basic Hash table set-up
HTcapacity = 98           # Allocate 97 slots in the Hash Table - can be varied
HT = [None]*HTcapacity    # 'HT' is the hash table data structure  
multFactor = 137          # multiplication factor used by Hash function - can be varied
print(len(HT))            # Verify HT capacity

fname = 'AS6input.txt'  # Make this the name of your local file
f = open(fname, 'r')
a6input = f.readlines()
f.close()
print(len(a6input))
print(a6input[0], '   ', a6input[-1])
print(a6input[80])

inlists = [ ]        # Create a list-of-lists for the input transactions
# insert code here to split the a6input lines into the list-of-lists
for line in a6input:
    transaction_line = line.strip().split(",")
    if transaction_line[0] =="I":  
        inlists.append(transaction_line)
    elif transaction_line[0] =="F":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            find_count += 1
    elif transaction_line[0] =="R":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            HT[hash_val].remove(transaction_line[1:])
            remove_count += 1

def Hash (inline, mFactor = 1123):
    """Hash function - takes an input line and returns the HT code. """
    # insert code to convert an input line string into a number that ranges from
    # 0 to the capacity of the hash table HT
    transact_line=''.join(inline)
    hash_num = 0
    for i in transact_line:
        hash_num = ord(i) * mFactor * 2   #Turns string into int
    return hash_num % 25            #Returns remainder as hash

# This runs all the input through the hash function and prints te resulting number
# Code is provided - no need to rewrite this cell.
for l in inlists:    # inlists is created in Step (3), has the input to be hashed
    if len(l) > 3:
        print('Hash values for "I" input records: ', l[1], l[2], ' is ',Hash(l, mFactor = multFactor))


# Recompute the has value of each line.  Store the entire input record in the HT 'list' slot.  Chain
#   collisions by making the chain a 'list-of'list'.
HT = [None] * HTcapacity         # Create the HT, even if done in Step (1)
for i in inlists:
    if i[0] == 'I':              # only handle insertions
    # insert code to calculate the hash and insert into the table
        hash_val = Hash(i, mFactor=multFactor)
    # also handle collisions - can use probing or chain by appending results into a 
    # that occupies the table slot
        if HT[hash_val] is None:
            HT[hash_val] = [i]
            print("Inserted into slot:", i)
        else:
            HT[hash_val].append(i)
            print("[Collision occured] Inserted into slot:", i)

# Print the hash table contents - shows what's in the table
# Code provided - no need to recreate
for n in range(0, len(HT)):
    print('index #', n, ' = ', HT[n])

empty = 0            # empty HT slots
occupied = 0         # HT slots occupied (not = None)
collisions = 0       # HT slots with collisions - more than 1 entry at the same place
items = 0            # entries in the table and collision chains
chains = 0           # entries in HT table slots with > 1 entry (i.e. entry is a collision slot)
# Iterate through the HT and calculate the vairous stats - some are above
# Code goes here...
for slot in HT:
    if slot is None:
        empty += 1
    else:
        occupied += 1
        if len(slot)>1:
            collisions += 1       #increment by 1 since len(slot)>1 = collision
            chains += len(slot)-1   #length of (chained) slot minus 1 (first element)
        items += len(slot)  #increment item if slot is occupied
load_factor = items / HTcapacity

#Hash table stats
print("Empty slots:", empty)
print("Occupied slots:", occupied)
print("Collisions:", collisions)
print("Total items:", items)
print("Chains:", chains)
print("Load Factor:", load_factor)


98
105
I,Rizzo,Mike,1962,M,PhEd
     F,Sagan,Carl,1938,M,Astronomy
I,Rodriguez,Hector,1987,M,ENGR

Hash values for "I" input records:  Rizzo Mike  is  0
Hash values for "I" input records:  Womack Mary  is  21
Hash values for "I" input records:  Smith Marv  is  17
Hash values for "I" input records:  Miller Samuel  is  16
Hash values for "I" input records:  Kelvin Katrina  is  3
Hash values for "I" input records:  Sherman Maria  is  16
Hash values for "I" input records:  Ali Hassan  is  8
Hash values for "I" input records:  Smith Mary  is  18
Hash values for "I" input records:  Miller Smuel  is  16
Hash values for "I" input records:  Adams Stuart  is  0
Hash values for "I" input records:  Lomack Marv  is  9
Hash values for "I" input records:  Strasburg Stephen  is  0
Hash values for "I" input records:  Thompson Creek  is  16
Hash values for "I" input records:  Bryson William  is  24
Hash values for "I" input records:  Eaton Adam  is  21
Hash values for "I" input records:  Jones Johann  i

#### Test #3 For this test, I changed the variable of the mFactor by double to 2246 instead of 1123.

In [67]:
# Basic Hash table set-up
HTcapacity = 98           # Allocate 97 slots in the Hash Table - can be varied
HT = [None]*HTcapacity    # 'HT' is the hash table data structure  
multFactor = 137          # multiplication factor used by Hash function - can be varied
print(len(HT))            # Verify HT capacity

fname = 'AS6input.txt'  # Make this the name of your local file
f = open(fname, 'r')
a6input = f.readlines()
f.close()
print(len(a6input))
print(a6input[0], '   ', a6input[-1])
print(a6input[80])

inlists = [ ]        # Create a list-of-lists for the input transactions
# insert code here to split the a6input lines into the list-of-lists
for line in a6input:
    transaction_line = line.strip().split(",")
    if transaction_line[0] =="I":  
        inlists.append(transaction_line)
    elif transaction_line[0] =="F":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            find_count += 1
    elif transaction_line[0] =="R":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            HT[hash_val].remove(transaction_line[1:])
            remove_count += 1

def Hash (inline, mFactor = 2246):
    """Hash function - takes an input line and returns the HT code. """
    # insert code to convert an input line string into a number that ranges from
    # 0 to the capacity of the hash table HT
    transact_line=''.join(inline)
    hash_num = 0
    for i in transact_line:
        hash_num += ord(i) * mFactor   #Turns string into int
    return hash_num % HTcapacity            #Returns remainder as hash

# This runs all the input through the hash function and prints te resulting number
# Code is provided - no need to rewrite this cell.
for l in inlists:    # inlists is created in Step (3), has the input to be hashed
    if len(l) > 3:
        print('Hash values for "I" input records: ', l[1], l[2], ' is ',Hash(l, mFactor = multFactor))


# Recompute the has value of each line.  Store the entire input record in the HT 'list' slot.  Chain
#   collisions by making the chain a 'list-of'list'.
HT = [None] * HTcapacity         # Create the HT, even if done in Step (1)
for i in inlists:
    if i[0] == 'I':              # only handle insertions
    # insert code to calculate the hash and insert into the table
        hash_val = Hash(i, mFactor=multFactor)
    # also handle collisions - can use probing or chain by appending results into a 
    # that occupies the table slot
        if HT[hash_val] is None:
            HT[hash_val] = [i]
            print("Inserted into slot:", i)
        else:
            HT[hash_val].append(i)
            print("[Collision occured] Inserted into slot:", i)

# Print the hash table contents - shows what's in the table
# Code provided - no need to recreate
for n in range(0, len(HT)):
    print('index #', n, ' = ', HT[n])

empty = 0            # empty HT slots
occupied = 0         # HT slots occupied (not = None)
collisions = 0       # HT slots with collisions - more than 1 entry at the same place
items = 0            # entries in the table and collision chains
chains = 0           # entries in HT table slots with > 1 entry (i.e. entry is a collision slot)
# Iterate through the HT and calculate the vairous stats - some are above
# Code goes here...
for slot in HT:
    if slot is None:
        empty += 1
    else:
        occupied += 1
        if len(slot)>1:
            collisions += 1       #increment by 1 since len(slot)>1 = collision
            chains += len(slot)-1   #length of (chained) slot minus 1 (first element)
        items += len(slot)  #increment item if slot is occupied
load_factor = items / HTcapacity


#Hash table stats
print("Empty slots:", empty)
print("Occupied slots:", occupied)
print("Collisions:", collisions)
print("Total items:", items)
print("Chains:", chains)
print("Load Factor:", load_factor)


98
105
I,Rizzo,Mike,1962,M,PhEd
     F,Sagan,Carl,1938,M,Astronomy
I,Rodriguez,Hector,1987,M,ENGR

Hash values for "I" input records:  Rizzo Mike  is  63
Hash values for "I" input records:  Womack Mary  is  19
Hash values for "I" input records:  Smith Marv  is  6
Hash values for "I" input records:  Miller Samuel  is  31
Hash values for "I" input records:  Kelvin Katrina  is  53
Hash values for "I" input records:  Sherman Maria  is  90
Hash values for "I" input records:  Ali Hassan  is  1
Hash values for "I" input records:  Smith Mary  is  35
Hash values for "I" input records:  Miller Smuel  is  50
Hash values for "I" input records:  Adams Stuart  is  19
Hash values for "I" input records:  Lomack Marv  is  1
Hash values for "I" input records:  Strasburg Stephen  is  9
Hash values for "I" input records:  Thompson Creek  is  38
Hash values for "I" input records:  Bryson William  is  8
Hash values for "I" input records:  Eaton Adam  is  45
Hash values for "I" input records:  Jones Johann  

#### Test #4 For this test, I tried changing the HTCapacity to half of 98 which is equal to 49 available slots in the Hash table:

In [56]:
# Basic Hash table set-up
HTcapacity = 49           # Allocate 97 slots in the Hash Table - can be varied
HT = [None]*HTcapacity    # 'HT' is the hash table data structure  
multFactor = 137         # multiplication factor used by Hash function - can be varied
print(len(HT))            # Verify HT capacity

fname = 'AS6input.txt'  # Make this the name of your local file
f = open(fname, 'r')
a6input = f.readlines()
f.close()
print(len(a6input))
print(a6input[0], '   ', a6input[-1])
print(a6input[80])

inlists = [ ]        # Create a list-of-lists for the input transactions
# insert code here to split the a6input lines into the list-of-lists
for line in a6input:
    transaction_line = line.strip().split(",")
    if transaction_line[0] =="I":  
        inlists.append(transaction_line)
    elif transaction_line[0] =="F":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            find_count += 1
    elif transaction_line[0] =="R":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            HT[hash_val].remove(transaction_line[1:])
            remove_count += 1

def Hash (inline, mFactor = 1123):
    """Hash function - takes an input line and returns the HT code. """
    # insert code to convert an input line string into a number that ranges from
    # 0 to the capacity of the hash table HT
    transact_line=''.join(inline)
    hash_num = 0
    for i in transact_line:
        hash_num += ord(i) * mFactor   #Turns string into int
    return hash_num % HTcapacity            #Returns remainder as hash

# This runs all the input through the hash function and prints te resulting number
# Code is provided - no need to rewrite this cell.
for l in inlists:    # inlists is created in Step (3), has the input to be hashed
    if len(l) > 3:
        print('Hash values for "I" input records: ', l[1], l[2], ' is ',Hash(l, mFactor = multFactor))


# Recompute the has value of each line.  Store the entire input record in the HT 'list' slot.  Chain
#   collisions by making the chain a 'list-of'list'.
HT = [None] * HTcapacity         # Create the HT, even if done in Step (1)
for i in inlists:
    if i[0] == 'I':              # only handle insertions
    # insert code to calculate the hash and insert into the table
        hash_val = Hash(i, mFactor=multFactor)
    # also handle collisions - can use probing or chain by appending results into a 
    # that occupies the table slot
        if HT[hash_val] is None:
            HT[hash_val] = [i]
            print("Inserted into slot:", i)
        else:
            HT[hash_val].append(i)
            print("[Collision occured] Inserted into slot:", i)

# Print the hash table contents - shows what's in the table
# Code provided - no need to recreate
for n in range(0, len(HT)):
    print('index #', n, ' = ', HT[n])

empty = 0            # empty HT slots
occupied = 0         # HT slots occupied (not = None)
collisions = 0       # HT slots with collisions - more than 1 entry at the same place
items = 0            # entries in the table and collision chains
chains = 0           # entries in HT table slots with > 1 entry (i.e. entry is a collision slot)
# Iterate through the HT and calculate the vairous stats - some are above
# Code goes here...
for slot in HT:
    if slot is None:
        empty += 1
    else:
        occupied += 1
        if len(slot)>1:
            collisions += 1       #increment by 1 since len(slot)>1 = collision
            chains += len(slot)-1   #length of (chained) slot minus 1 (first element)
        items += len(slot)  #increment item if slot is occupied
load_factor = items / HTcapacity


#Hash table stats
print("Empty slots:", empty)
print("Occupied slots:", occupied)
print("Collisions:", collisions)
print("Total items:", items)
print("Chains:", chains)
print("Load Factor:", load_factor)

49
105
I,Rizzo,Mike,1962,M,PhEd
     F,Sagan,Carl,1938,M,Astronomy
I,Rodriguez,Hector,1987,M,ENGR

Hash values for "I" input records:  Rizzo Mike  is  14
Hash values for "I" input records:  Womack Mary  is  19
Hash values for "I" input records:  Smith Marv  is  6
Hash values for "I" input records:  Miller Samuel  is  31
Hash values for "I" input records:  Kelvin Katrina  is  4
Hash values for "I" input records:  Sherman Maria  is  41
Hash values for "I" input records:  Ali Hassan  is  1
Hash values for "I" input records:  Smith Mary  is  35
Hash values for "I" input records:  Miller Smuel  is  1
Hash values for "I" input records:  Adams Stuart  is  19
Hash values for "I" input records:  Lomack Marv  is  1
Hash values for "I" input records:  Strasburg Stephen  is  9
Hash values for "I" input records:  Thompson Creek  is  38
Hash values for "I" input records:  Bryson William  is  8
Hash values for "I" input records:  Eaton Adam  is  45
Hash values for "I" input records:  Jones Johann  is

#### Test #5 Changing the HTCapacity to 200

In [62]:
# Basic Hash table set-up
HTcapacity = 200           # Allocate 97 slots in the Hash Table - can be varied
HT = [None]*HTcapacity    # 'HT' is the hash table data structure  
multFactor = 137         # multiplication factor used by Hash function - can be varied
print(len(HT))            # Verify HT capacity

fname = 'AS6input.txt'  # Make this the name of your local file
f = open(fname, 'r')
a6input = f.readlines()
f.close()
print(len(a6input))
print(a6input[0], '   ', a6input[-1])
print(a6input[80])

inlists = [ ]        # Create a list-of-lists for the input transactions
# insert code here to split the a6input lines into the list-of-lists
for line in a6input:
    transaction_line = line.strip().split(",")
    if transaction_line[0] =="I":  
        inlists.append(transaction_line)
    elif transaction_line[0] =="F":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            find_count += 1
    elif transaction_line[0] =="R":
        hash_val = Hash(transaction_line[1:], mFactor=multFactor)
        if HT[hash_val] is not None and transaction_line[1:] in HT[hash_val]:
            HT[hash_val].remove(transaction_line[1:])
            remove_count += 1

def Hash (inline, mFactor = 1123):
    """Hash function - takes an input line and returns the HT code. """
    # insert code to convert an input line string into a number that ranges from
    # 0 to the capacity of the hash table HT
    transact_line=''.join(inline)
    hash_num = 0
    for i in transact_line:
        hash_num += ord(i) * mFactor   #Turns string into int
    return hash_num % HTcapacity            #Returns remainder as hash

# This runs all the input through the hash function and prints te resulting number
# Code is provided - no need to rewrite this cell.
for l in inlists:    # inlists is created in Step (3), has the input to be hashed
    if len(l) > 3:
        print('Hash values for "I" input records: ', l[1], l[2], ' is ',Hash(l, mFactor = multFactor))


# Recompute the has value of each line.  Store the entire input record in the HT 'list' slot.  Chain
#   collisions by making the chain a 'list-of'list'.
HT = [None] * HTcapacity         # Create the HT, even if done in Step (1)
for i in inlists:
    if i[0] == 'I':              # only handle insertions
    # insert code to calculate the hash and insert into the table
        hash_val = Hash(i, mFactor=multFactor)
    # also handle collisions - can use probing or chain by appending results into a 
    # that occupies the table slot
        if HT[hash_val] is None:
            HT[hash_val] = [i]
            print("Inserted into slot:", i)
        else:
            HT[hash_val].append(i)
            print("[Collision occured] Inserted into slot:", i)

# Print the hash table contents - shows what's in the table
# Code provided - no need to recreate
for n in range(0, len(HT)):
    print('index #', n, ' = ', HT[n])

empty = 0            # empty HT slots
occupied = 0         # HT slots occupied (not = None)
collisions = 0       # HT slots with collisions - more than 1 entry at the same place
items = 0            # entries in the table and collision chains
chains = 0           # entries in HT table slots with > 1 entry (i.e. entry is a collision slot)
# Iterate through the HT and calculate the vairous stats - some are above
# Code goes here...
for slot in HT:
    if slot is None:
        empty += 1
    else:
        occupied += 1
        if len(slot)>1:
            collisions += 1       #increment by 1 since len(slot)>1 = collision
            chains += len(slot)-1   #length of (chained) slot minus 1 (first element)
        items += len(slot)  #increment item if slot is occupied
load_factor = items / HTcapacity


#Hash table stats
print("Empty slots:", empty)
print("Occupied slots:", occupied)
print("Collisions:", collisions)
print("Total items:", items)
print("Chains:", chains)
print("Load Factor:", load_factor)

200
105
I,Rizzo,Mike,1962,M,PhEd
     F,Sagan,Carl,1938,M,Astronomy
I,Rodriguez,Hector,1987,M,ENGR

Hash values for "I" input records:  Rizzo Mike  is  165
Hash values for "I" input records:  Womack Mary  is  79
Hash values for "I" input records:  Smith Marv  is  80
Hash values for "I" input records:  Miller Samuel  is  111
Hash values for "I" input records:  Kelvin Katrina  is  93
Hash values for "I" input records:  Sherman Maria  is  148
Hash values for "I" input records:  Ali Hassan  is  131
Hash values for "I" input records:  Smith Mary  is  93
Hash values for "I" input records:  Miller Smuel  is  96
Hash values for "I" input records:  Adams Stuart  is  105
Hash values for "I" input records:  Lomack Marv  is  157
Hash values for "I" input records:  Strasburg Stephen  is  85
Hash values for "I" input records:  Thompson Creek  is  142
Hash values for "I" input records:  Bryson William  is  66
Hash values for "I" input records:  Eaton Adam  is  69
Hash values for "I" input records:  J